### ISSUE TITLE
[TW-230] Filter tasks on partial UUIDs

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#274]
_Paul Kishimoto on 2014-01-10T22:10:20Z says:_ 

 (at)git@ also identifies its basic units (commits) with hashes:

<pre>
$ git log -1
commit 97f74b72765a15fa050cbfa442e7dc944e4028bc
Author: Paul Natsuo Kishimoto <mail(at)paul.kishimoto.name>
Date:   Fri Dec 13 17:52:49 2013 -0500

    New script: kdx
</pre>

The first few characters of the hash can also be used to identify the commit on the command line:
<pre>
$ git log -1 97f7                                                                                           
commit 97f74b72765a15fa050cbfa442e7dc944e4028bc
Author: Paul Natsuo Kishimoto <mail(at)paul.kishimoto.name>
Date:   Fri Dec 13 17:52:49 2013 -0500

    New script: kdx
</pre>
…but only if unambiguous:
<pre>
$ git log | grep "commit 9"
commit 97f74b72765a15fa050cbfa442e7dc944e4028bc
commit 97308807312feb209deff8d00859b59d6a889902
commit 9d30c0c30a60ed0e8f730bb3b1c2f42fecf84e34
$ git log -1 9
fatal: ambiguous argument '9': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
</pre>

It would be nice if this were possible with task UUIDs—when attempting to select completed tasks, it would eliminate the need to either copy/paste or fully retype a UUID read from output of a previous command.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2014-01-10T22:10:20Z
Modified: 2014-07-02T03:52:04Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-01-11T00:09:48Z says:_ 

 You can already use partial UUIDs, but 14 characters is the minimum.  This is because any fewer causes havoc with the unit tests, and therefore probably users.

Git has the luxury of accepting fewer alternatives on the command line - it's either a label or a treeish SHA1.  Taskwarrior has a more complex task, for example, suppose you have only two tasks, id 1 (uuid 20000000-0000-0000-0000-000000000000) and id 2 (uuid 10000000-0000-0000-0000-000000000000).  Which one should the following command select?

<pre>
task 1 delete
</pre>

So unambiguous UUIDs can collide with IDs and the result is not good.  You have to read a lot of characters (depending on the UUID value) in order to determine conclusively that it is not a date (20140110, 201401, 1234), an id (1, 12, 123, 1234, 12345) nor anything else.

**Comment by taskwarrior** [comment]
_Paul Kishimoto on 2014-01-12T19:17:17Z says:_ 

 In git, AFAIK, tags and branches can be given names that begin with a decimal digit (0–9), so "unambiguous" does mean that the hash fragment not only identifies a single hash, but also that it is neither a tag nor a branch name.

But I do appreciate that implementing this seemingly-simple feature is non-trivial.

I was going to point out some helpful things (for instance, something like 76% of UUIDs will have an a-f among the first 3 digits, 85% in the first 4, 90% in the first 5, etc.)—but I realize I don't know how the filter code works, so these might not simplify or speed up an implementation.

If adding a lot of such logic is onerous or would impact performance, even the ability to match an initial UUID fragment with "(at)task uuid:d72@" would be an improvement.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-01-12T20:33:02Z says:_ 

 > the ability to match an initial UUID fragment with "(at)task uuid:d72@" would be an improvement.

Actually, that's already supported:

<pre>
task uuid.startswith:d72 info
</pre>

But of course, whether that is ambiguous or not is another matter.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-05-13T04:48:41Z says:_ 

 This will be implemented in 2.4.0, at least in spirit, using the partial match operator.
